### PR TITLE
Do not allow array modifications after freeze inside sort!

### PR DESCRIPTION
**Pull request by jeremyevans**
If freezing an array inside sort!, previously the array could be
modified after the freeze.  This checks whether the receiver is
frozen after every yield, preventing modifications if the receiver
is frozen inside the block.

Fixes [Bug #17739]

### CODE REVIEW

File: array.c [hunk 1]
```diff
@@ -3232,6 +3233,9 @@ sort_1(const void *ap, const void *bp, void *dummy)
     args[0] = a;
     args[1] = b;
     retval = rb_yield_values2(2, args);
+    if (rb_obj_frozen_p(data->receiver)) {
+        rb_raise(rb_eFrozenError, "array frozen during sort");
+    }
     n = rb_cmpint(retval, a, b);
```
Comment (nobu) [inline array.c:3239]: Also `retval.<` or `retval.>` in `rb_cmpint` can freeze the receiver.
And `sort_2` needs this check too.

Comment (jeremyevans) [reply array.c]: @nobu  Thank you for your review.  I have rebased this on master and fixed the two cases you mentioned.

Comment (nobu) [reply array.c]: Maybe extract a function with `sort_reentered`?

Comment (jeremyevans) [reply array.c]: That makes sense.  I've pushed another commit that adds a `sort_returned` function that calls `sort_reentered` and has the frozen check.

**Pull request merged**